move logic for manipulating strides to ViewBuilder.

rainwoodman opened the pull request.

rainwoodman committed:
Move stride shape manipulation to ViewBuilder.

rainwoodman committed:
add example slicing with view builder.

arteymix commented on tests/vast-test.vala:
```diff
@@ -209,6 +209,31 @@ int main (string[] args) {
+            }
+        }
+
+        var b = a.view().slice(0, -1, 5, ssize_t.MAX).finish();
```
Could you rename `finish` for `apply`? or `end`?

rainwoodman committed:
add example rerouting axis.

arteymix commented:
I wonder whether we should pass the axis to `slice`. Maybe have a counter everytime we slice?

```vala
arr.view ()
   .slice (0, size_t.MAX) // axis 0
   .slice (0, size_t.MAX) // axis 1
   .skip ()     // skip axis 2
   .end ();
```

rainwoodman committed:
add broadcast support.

rainwoodman commented:
No. not a counter -- they are intended to be stateless -- because we need to be able to call this api in a loop or selectively apply them on some axes.

rainwoodman committed:
rename finish to end.

rainwoodman committed:
rename ssize_t.MAX to THRU

or we need a better name for this magic number.

rainwoodman committed:
give magic number names.

rainwoodman committed:
add example using Array.THRU.

rainwoodman committed:
add notes, extend broadcastable to strides[axis] == 0

arteymix commented:
Then maybe a `axis` selector that would use the default axis if not specified:

```vala
a.view ()
    .axis (0).slice (1, 5)
    .axis (1).slice (4, 9)
    .end ();
```

Let's say that a buch of number is hardly readable for a builder.

arteymix commented:
To remain consistent with the other APIs, we could have:

 - slice (from, to, step = 1)
 - step (step)
 - head (to)
 - tail (from)
 - swap (other_axis)

rainwoodman commented:
Indeed it is hardly readable, but vala doesn't support named arguments -- it is a defect in the language. 

I strongly prefer not having selectors -- it adds in a state to the ViewBuilder and thus reminds me cin / cout of C++. The formatter are uttly verbose. 

The StringBuilder in vala is much nicer in comparison.

rainwoodman commented:
The idea of the ViewBuilder is to have a mechanism to implement higher level operations in the array, not to implement the higher level operations internally, so these methods shall stay in Array as convienent wrappers to .view(), in my opinion.

rainwoodman commented:
I originally put step before slice because the default value of from and to are realized using the value of step. I think the tradition is having step later. My brain short-circuited. I'll switch the order.

rainwoodman committed:
use the human friendly slice(axis, from, to, step) ordering.

arteymix commented:
I agree on the necessity of such feature: it's not convenient to work on individual axis otherwise.

Maybe at least introduce `head` and `tail` to avoid using arbitrary constants?

rainwoodman commented:
To be complete, we'll need four functions to cover each of these:

```
a:b:c (slice)
:b:c (tail)
::c (step)
a::c (head)
```
(because default value of c is 1, in the domain of ssize_t)

I really would rather having just one function... Do you have a reference removing ssize_t.MAX from the domain causes problem in real world?

rainwoodman commented:
I sense that as long as we don't poke a hole (thus we shall remove from the edge of the domain, e.g. ssize_t.MAX) it won't be mathmatically uglier than the original definition of ssize_t.

arteymix commented:
I feel like it's a very dirty hack. Normally, that would cause an out-of-bound array access.

We're not in Python here, so we have to be consequent.

Also, I don't like the word `View`: our arrays are basically all views over a `GBytes` object. Thus, I think `Array.Builder` would suffice.

Sorry for being so negative, I only mean best for this project and designing good APIs is important.

rainwoodman commented:
I have a uglier proposal
what about
```
.slice(0, null, {0}, {1})
```
null is THRU. problem solved.

rainwoodman committed:
rename ViewBuilder to builder.

rainwoodman committed:
rename view -> build

arteymix commented:
Now C users have to declare in advance a size_t* pointer before passing it. Honestly, stop resisting! Python bindings will have the nice slicing protocol anyway!

arteymix commented:
Can we have a multi-dimensional broadcast utility as well? We might also need a way of finding a broadcasting shape for multiple arrays, possibly into `MultiIterator`.

rainwoodman commented:
OK -- four functions on the go. 

That way of finding shape shall be somewhere else -- not here then a follow up function makes use of Builder to create broadcasted views.

arteymix commented:
I'm going to sleep. I'll review all this tomorrow if you want. It sounds really good and elegant to me.

rainwoodman committed:
add head tail step.

These are really really more abstract than triplets.

rainwoodman committed:
Add qslice that uses pointer promotion dealing with undefined numbers.

rainwoodman commented:
OK. I added a qslice that covers all four cases with pointer promotion. There is a lot of code duplication and I do prefer *not* to refactor them into really short functions -- not for performance but for readability.

rainwoodman committed:
change head tail slice step to wrappers of qslice.

rainwoodman commented:
Ok, the best of both worlds. we have five functions.
(head, tail, step, slice -- for c binding) all calls qslice (internal, vala, python, ...).

rainwoodman committed:
remove THRU

arteymix commented:
For `qslice`, just using `ssize_t?` is sufficient to promote the type to a pointer.

It really looks nice to me :+1:.

What about reverse slicing? The reason I added it in `Array.slice` is that it was not dealing with `step` as well, so it made sense to use negative step if the slice is reversed.

arteymix commented:
`qslice` could be named `slice`, but use a different `cname`. This way we would have both `slice` for C and Vala but with slightly different semantics, yet an uniform API.

arteymix commented on src/array.vala:
```diff
@@ -464,4 +396,175 @@ public class Vast.Array : Object
+            ssize_t []? qfrom = null,
+            [CCode (array_length = false)]
+            ssize_t []? qto = null,
+            ssize_t step = 1)
```
All the logic should be in `slice`, `qslice` would only compute default values when `null` is given.

rainwoodman replied:
There would be a lot of duplicated logic if we do it that way and do it right. We will need a internal method that does unprotected slicing -- because slice would wrap the dimension and mess it up (when `to` is -1).

In the old version ( 457a4e4 ) I had a `_slice` method that does unprotected slicing and everyone preprocess the defaults and calls `_slice`. You can see the code is a lot fatter than this version.

arteymix commented on src/array.vala:
```diff
@@ -464,4 +396,175 @@ public class Vast.Array : Object
+        }
+
+        public Builder
+        reset_dimension(size_t dimension)
```
Is this really useful? Maybe have a `revert` method to return the original array instead?

```vala
arr.build ()
    .slice (0, ...)
    .revert ();
```

rainwoodman replied:
I don't think this is useful. I'll remove it.

arteymix commented on src/array.vala:
```diff
@@ -464,4 +396,175 @@ public class Vast.Array : Object
+            return new Array (array.scalar_type,
+                              array.scalar_size,
+                              shape[0:dimension],
+                              strides[0:dimension],
```
No need to slice the `strides` here, the array does not have a length.

rainwoodman replied:
I did it because it looks symmetric.

arteymix commented on tests/vast-test.vala:
```diff
@@ -237,8 +311,9 @@ int main (string[] args) {
-        assert (-1 * sizeof (int64) == a.slice ({10, 10}, {0, 0}).strides[1]);
+        // this shall die, slice cannot handle this because it assumes step == 1
+        // assert (10 == a.slice ({10, 10}, {0, 0}).shape[1]);
+        // assert (-1 * sizeof (int64) == a.slice ({10, 10}, {0, 0}).strides[1]);
```
Thus no reverse slicing? We might want to add `step = {}` to the `Array.slice` signature then.

rainwoodman replied:
There is reversed slicing the the slice method of array doesn't support it -- it always does step == 1. 

In Python if you set from >= to without explicitly say step=-1 you get an empty array.

I think we need to follow that convention -- but because we insisted this slice function only describes `a:b:`, there is no way to pass in a step.

rainwoodman replied:
Yes I think we shall add step = {}. the default is null.

arteymix commented:
That said, if we want Vala-specific APIs, we're better adding a `vast-1.0-custom.vapi` and put the code literally there. It will avoid polluting the symbol space with unusable functions from C.

arteymix commented on src/array.vala:
```diff
@@ -2,6 +2,9 @@ using GLib;
 public class Vast.Array : Object
 {
+    public const ssize_t NEW_AXIS = ssize_t.MAX;
+    public const size_t DEFAULT_DIMENSION = size_t.MAX;
```
Maybe we could have a view that extends the dimension instead of having a default dimension?

```vala
var arr = new Vast.Array (..., {2, 2, 2, 2});
arr.redim (5).build ().slice (4, 0, 1);
```

That would basically call `reshape` with paddings of `1`.

arteymix replied:
Ok, it's here: a50a23b582202d0159b9bb858fa70940c8b6d506

arteymix replied:
For `NEW_AXIS` if we are to add an axis, it's just a matter of:

```
arr.redim (arr.dimension + 1);
```

This way, we don't really need any constants ;)

rainwoodman replied:
Not its insufficient -- what if I want a new axis in the middle? e.g. `[:, Newaxis, :]`?

rainwoodman replied:
We may need to keep track of some of the `axis` changes to ensure no original axes is used twice, because

the current way we can end up with broken strides and shape. e.g.

```
.axis(0, 0)
.axis(1, 0)
```

arteymix replied:
Just do `arr.redim (5).transpose ({0, 1, 4, 2, 3})`. Basically add new axis and transpose it to the middle. We might have a helper for inserting an axis or redim at a given position.

```vala
arr.redim (5, 3); // insert new axis after dimension 3
```

arteymix replied:
I think it's more intuitive to insert axis before than after.

rainwoodman replied:
First, there is no insertion -- `axis` assigns / routes the axis to an axis of the original array or a newaxis. The dimension of the builder is already set when it is created. (`.builder(5)`).

I have been considering using transpose too -- a few drawbacks such that I didn't do it:

- The equivalent of `[:, newaxis, ...]` would be `arr.redim(arr.dimension + 1).transpose({0, dimension + 1, 1, 2, ...})`. It is a lot less direct -- I wanted to insert an axis but I have to do it via a transpose -- and create a full permutation array whatever I want to do?


- On the other hand, implementing `transpose` with `axis` is easy.

- In cases I want to swap two axis, `.transpose({1, 0, 2, 3, 4 ...})` I need to create redundant entries in the argument. compare this to current way, `.axis(0, 1).axis(1, 0)`. 

- thus I concluded, routining/assignment of axis (`axis`) is more fundamental than `.transpose`.

- I don't think the current function name `axis` correctly describes what the function does.

rainwoodman replied:
The drawback of `axis` is that in cases the axis assignment is non-sense -- hence I added sanity check in `end()` to make sure each original axis is at most used once. (I think we can assert it is used exactly once).

arteymix commented:
I'm really glad of these changes, it will make axis manipulation a breeze ;) I don't think NumPy has anything similar.

rainwoodman commented:
oh man conflict...

rainwoodman committed:
squash reset_dimensions -- not useful.

rainwoodman committed:
check each original axis is used at most once in the new array.

rainwoodman committed:
add .index and step == 0 for reducing the dimension of array

This selects a single element along that axis and removes the axis
corresponds to indexing with mixed scalar and slice in Python.
.e.g.  [3, 1:3, :]

rainwoodman committed:
use vala's nullable ssize_t.

what a breeze in the test.

rainwoodman commented:
The nullable ssize_t is neat look at the test case -- do we still need the head tail methods?

The original concern regarding requiring a pointer syntax in C is not that bad -- most of the cases the parameters are passed in as an addressable variable, not a number literal anyways. Ocacionally we do need use literals there must be a macro to make it addressable, e.g., in c99,

```
#define _A(a) (gssize_t []) {a}
```

arteymix commented on tests/vast-test.vala:
```diff
@@ -313,6 +298,28 @@ int main (string[] args) {
+        assert (f1.get_value({1, 2}).get_double() == 2 * 7);
+
+        var g = a.build()
+                 .index(1, 5)
```
Just to be sure, is this conceptually equivalent to the `Array.index` utility I introduced?

https://github.com/rainwoodman/vast/blob/master/src/array.vala#L251

rainwoodman replied:
It is the same kind of operation, but this one allows you to index from any picked axis. Array.index only allows the first-n dimensions.

arteymix commented:
I still prefer `head` and `tail` though because it does not involve any trick. If you are really to slice from and to an edge, it's better being explicit, no?

Maybe we can try to patch Vala slice protocol to include `[a:]` and `[:b]` notation? and also stepping?

arteymix commented:
If you want I can merge the PR and deal with conflicts. I think I'll have to rewrite a few utilities I added recently, but that's okay.

rainwoodman commented:
On the other hand, I was already able to make mistakes translating the slice to head/tail. 

I pondered how and realized confusion is that 

- with `head` you specify the 'end'
- with `tail` you specify the `begin`

So it is fundamentally a pair of mind twisters -- take time to get used to and may come back and bite you later.

rainwoodman commented:
yes. Let's merge this and talk about vala patch later.

arteymix commented:
Okay, good, I'll merge this locally then.

arteymix commented:
Ok, it seems like you've already rebased everything. I leave you the honor to merge!
